cmd/linode-cosi-driver/main.go

```diff
@@ -196,3 +242,93 @@ func shutdown(ctx context.Context, wg *sync.WaitGroup, g *grpc.Server) {
 		}
 	}
 }
+
+func setupObservabillity(ctx context.Context,
+	tracesProtocol string,
+	metricsProtocol string,
+	userAgent string,
+) func() {
+	node := os.Getenv(envK8sNodeName)
+	pod := os.Getenv(envK8sPodName)
+
+	res := resource.NewWithAttributes(
+		semconv.SchemaURL,
+		semconv.ServiceName(driverName),
+		semconv.ServiceVersion(version.Version),
+		semconv.K8SPodName(pod),
+		semconv.K8SNodeName(node),
+		semconv.UserAgentOriginal(userAgent),
+	)
+
+	log.Info("setting up tracing",
+		"protocol", tracesProtocol)
+
+	tracingShutdown, err := tracing.Setup(ctx, res, tracesProtocol)
+	if err != nil {
+		// non critical error, just log it.
+		log.Error("failed to setup tracing",
+			"error", err,
+		)
+	}
+
+	log.Info("setting up metrics",
+		"protocol", metricsProtocol)
+
+	metricsShutdown, err := metrics.Setup(ctx, res, metricsProtocol)
+	if err != nil {
+		// non critical error, just log it.
+		log.Error("failed to setup metrics",
+			"error", err,
+		)
+	}
+
+	attrs := []any{}
+
+	for _, kv := range os.Environ() {
+		k, v, ok := strings.Cut(kv, "=")
+		if ok && strings.HasPrefix(k, "OTEL_") {
+			attrs = append(attrs, slog.String(k, v))
+		}
+	}
+
+	log.Info("opentelemetry configuration applied",
+		attrs...,
+	)
+
+	return func() {
+		ctx, cancel := context.WithTimeout(context.WithoutCancel(ctx), time.Second)
```

Comment:
I've no idea how long we'd expect it to take to flush the remaining spans, do you think 1s is long enough? If we're defaulting to talking gRPC to a sidecar I guess 1s is quite long in computer terms.

Reply:
Good catch. With default OTLP timeout being 10000ms[^1] we should probably expect the final flushing timeout to be at least equal to the default timeout, and probably 2.5x that, so the exporters can retry the data upload.

[^1]: https://opentelemetry.io/docs/specs/otel/protocol/exporter/#:~:text=OTEL_EXPORTER_OTLP_TRACES_COMPRESSION%20OTEL_EXPORTER_OTLP_METRICS_COMPRESSION%20OTEL_EXPORTER_OTLP_LOGS_COMPRESSION-,Timeout,-%3A%20Maximum%20time%20the

Reply:
Ok if we set it to 25s then when we add signal handling we'll still be inside the default 30s `terminationGracePeriodSeconds` so that seems fine to me